src/hooks/useQueryState.js

```diff
@@ -0,0 +1,163 @@
+ * arrow, preventing you from going forward (and other similar browsing history issues).
+ *
+ * Current Implementation Issues:
+ * - UseLayoutEffect is not a perfect solution to the race condition issue. Symptoms of this
+ * possibly poor design are evident when the browser 'back'/'forward' arrows are tested heavily
+ * which can skip queryStates (likely not a practical use case), as well as the fact that
+ * we cannot set to an initial value or else we might load (and overwrite the state) from the url
+ * before we can set the url
+ *
+ * @param {String} keyParam - key of the QueryState variable
+ * @param {String} typeString - the type of variable (must correspond with a types enum value)
+ * @param {String} initial - initial value of the QueryState variable (optional)
+ *          *** DOES NOT currently work because of loading from URL before URL gets set
+ * @returns {[var, Callback]} - the state variable, the setQueryState callback function
+ */
+function useQueryState(keyParam, typeString, initial) {
+  const key = keyParam; // we should never change the particular key once we start using it
+  const types = {
+    SingleValue: 'SingleValue',
+    Array: 'Array',
+    Boolean: 'Boolean',
+  };
+  const defaultValues = {
+    SingleValue: '',
+    Array: [],
+    Boolean: false,
+  };
+  const type = types[typeString];
+  if (type === undefined)
+    throw new Error(`type must be one of the given types: ${Object.values(types)}`);
+  const defaultVal = defaultValues[type];
+  const [state, setState] = useState(initial ?? defaultVal);
+  const history = useHistory();
+
+  /**
+   * Anytime the state changes, setURLParam is invoked
+   * This function updates the URL with the state (if changed)
+   */
+  useEffect(() => {
+    const setURLParam = async () => {
+      let urlParams = new URLSearchParams(history.location.search);
+
+      // if the url value is already the same as the state value -> don't set the url again
+      // note: must include toString in defaultVal.toString() in order to check empty arrays
+      if (
+        (urlParams.get(key) ?? defaultVal.toString()) === state.toString() ||
+        (state === true && urlParams.get(key) === '')
+      )
+        return;
+
+      // key with no value OR boolean key set to false -> absent from url, push empty
+      if (!state || state === '' || state.length === 0 || state === false) urlParams.delete(key);
+      // boolean or value key is present in url with its current value
+      else urlParams.set(key, state.toString());
+
+      // boolean keys set to true don't need '=value'
+      urlParams = urlParams
+        .toString()
+        .replace(/=true|=$/g, '')
+        .replace(/=&/, '&');
+
+      history.push('?' + urlParams);
+    };
+    setURLParam();
+    /*********************** ESLint Disable Justification: *******************************/
+    // ESLint wants to include `defaultVal, key, history` as dependencies but to prevent
```

Comment:
Presumably, including `defaultVal` and `key` would not do any harm, since those are stable for the lifetime of the hook. Including them would also make it more clear that it's _specifically_ `history` that we do not want to depend on.